### DISCUSSION TITLE
Why do we have disposables anyway?

### DISCUSSION THREAD

**Discussion by daixtrose** [victimsnino/ReactivePlusPlus#680] [General]
Triggered by https://github.com/victimsnino/ReactivePlusPlus/pull/679, I feel inclined to comment that I do not understand why a reactive extensions implementation in **C++** requires the notion of a `disposable.` The [RAII](https://en.cppreference.com/w/cpp/language/raii) principle is there before your nose! Every piece of C++ software I saw in the past 30 years that put an extra layer for lifetime or resource management on top of RAII was doomed to fail. 

Can you elaborate on why we must copy a `C#`-feature here?

**Comment by AlexInLog** [comment]
disposable is actually reactive raii =) you can't just keep real raii object due to it is callbacks
Ok, to prepare better comment with explanation for disposables, could you explain your point of view and preferred solution to understand your initial thoughts =)

**Comment by daixtrose** [comment]
Everywhere you return a disposable, you could return a handle that, at the end of its lifetime, does the things that are currently done in `dispose()`. I simply do not want to call `dispose()`anywhere in C++.

**Reply by AlexInLog** [reply]
Yeah, it actually does it https://github.com/victimsnino/ReactivePlusPlus/blob/v2/src/rpp/rpp/disposables/disposable_wrapper.hpp#L45
comparing to rxcpp for example.
But sometimes end of lifetime could be a bit to long so rpp does manual dispose of disposables itself. I'm not expecting from you as user to call dispose anywhere (except for `subscribe_with_disposable` where you can manually decide to stop current observable)

**Reply by daixtrose** [reply]
OK, does that mean you are not expecting the user to call `dispose()` at all?

**Reply by AlexInLog** [reply]
Yes. Only if **YOU** **want** to dispose disposable **right now**, for example, to make early unsubscribe of observer or stop scheduled action.

**Reply by AlexInLog** [reply]
Looks like i have to mention it somewhere in docs exactly =)

**Reply by daixtrose** [reply]
I had hoped to obtain a `subscription` object when I call `subscribe` which in its desctructor would unsubscribe.

**Reply by AlexInLog** [reply]
Actually it is incorrect expectation. for example
```cpp
observable.subscribe(...); 
```
ooops, it causes immediate unsubscribe! Let's image i'm user and i'm dont care about subscription object: i don't want to do early unsubscribe, just subscribe. It would be missuse in this case..

i can provide raii object to support your case, but it shouldn't be default behavior
